src/lib.rs

```diff
@@ -339,10 +350,22 @@ impl<T: Contribution> Stream for WireMessages<T> {
     fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
         match try_ready!(self.framed.poll()) {
             Some(frame) => {
-                Ok(Async::Ready(Some(
-                    // deserialize_from(frame.reader()).map_err(Error::Serde)?
-                    bincode::deserialize(&frame.freeze()).map_err(Error::Serde)?,
-                )))
+                let s_msg: SignedWireMessage =
+                    bincode::deserialize(&frame.freeze()).map_err(Error::Serde)?;
+                let msg: WireMessage<T> =
+                    bincode::deserialize(&s_msg.message).map_err(Error::Serde)?;
+                if let WireMessageKind::HelloRequestChangeAdd(_, _, pk) = msg.kind {
+                    if self.their_key.is_some() {
+                        return Err(Error::DuplicateHello);
+                    }
+                    self.their_key = Some(pk);
+                }
+                if !self.their_key.as_ref()
+                        .map_or(true, |pk| pk.verify(&s_msg.sig, &s_msg.message)) {
+                    return Err(Error::InvalidSignature);
+                }
```

Comment:
This additional processing should probably take place within `Handler` somewhere (where node specific info is available).